File: 3544.subtree-inversion-sum.cpp
#
# @lc app=leetcode id=3544 lang=cpp
#
# [3544] Subtree Inversion Sum
#
# @lc code=start
class Solution {
public:
    long long subtreeInversionSum(vector<vector<int>>& edges, vector<int>& nums, int k) {
        int n = nums.size();
        vector<vector<int>> adj(n);
        for (auto& e : edges) {
            adj[e[0]].push_back(e[1]);
            adj[e[1]].push_back(e[0]);
        }
        
        map<tuple<int,int,int>, long long> memo;
        
        function<long long(int, int, int, int)> dfs = [&](int u, int parent, int d, int p) -> long long {
            auto key = make_tuple(u, d, p);
            if (memo.count(key)) return memo[key];
            
            int sign = (p == 0) ? 1 : -1;
            long long result = LLONG_MIN;
            
            vector<int> children;
            for (int v : adj[u]) {
                if (v != parent) {
                    children.push_back(v);
                }
            }
            
            // Option 1: Don't invert at u
            {
                long long sum = (long long)nums[u] * sign;
                int child_d = min(d + 1, k);
                for (int v : children) {
                    sum += dfs(v, u, child_d, p);
                }
                result = max(result, sum);
            }
            
            // Option 2: Invert at u (only if d >= k)
            if (d >= k) {
                long long sum = (long long)nums[u] * (-sign);
                for (int v : children) {
                    sum += dfs(v, u, 1, 1 - p);
                }
                result = max(result, sum);
            }
            
            return memo[key] = result;
        };
        
        return dfs(0, -1, k, 0);
    }
};
# @lc code=end

File: 3544.subtree-inversion-sum.py
#
# @lc app=leetcode id=3544 lang=python3
#
# [3544] Subtree Inversion Sum
#

# @lc code=start
class Solution:
    def subtreeInversionSum(self, edges: List[List[int]], nums: List[int], k: int) -> int:
        n = len(nums)
        adj = [[] for _ in range(n)]
        for u, v in edges:
            adj[u].append(v)
            adj[v].append(u)
        import sys
        sys.setrecursionlimit(50010)
        from functools import lru_cache
        MIN_VAL = - (n * 50001)
        @lru_cache(maxsize=None)
        def dfs(u: int, p: int, par: int, d: int) -> int:
            res = MIN_VAL
            for flip in [0, 1]:
                if flip == 1 and d < k:
                    continue
                sign = 1 if (par + flip) % 2 == 0 else -1
                contrib = nums[u] * sign
                cpar = (par + flip) % 2
                cd = 1 if flip else min(d + 1, k)
                tot_child = 0
                for v in adj[u]:
                    if v != p:
                        tot_child += dfs(v, u, cpar, cd)
                res = max(res, contrib + tot_child)
            return res
        return dfs(0, -1, 0, k)

# @lc code=end